src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +46,366 @@ public static String newChallenge(
+          weight += signerWithWeight.getWeight();
+          break;
+        }
+      }
+    }
+
+    if (weight < threshold) {
+      throw new InvalidSep10ChallengeException(String.format("Signers with weight %d do not meet threshold %d.", weight, threshold));
+    }
+
+    return signersFound;
+  }
+
+  private static List<String> verifyTransactionSignatures(Transaction transaction, List<String> signers) throws InvalidSep10ChallengeException {
+    List<DecoratedSignature> signatures = transaction.getSignatures();
+    if (signatures.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction has no signatures.");
+    }
+    List<String> signersFound = new ArrayList<String>();
+    for (String signer : signers) {
+      if (verifyTransactionSignedByAccountId(transaction, signer)) {
+        signersFound.add(signer);
+      }
+    }
+    return signersFound;
+  }
+
+  private static boolean verifyTransactionSignedByAccountId(Transaction transaction, String accountId) {
+    KeyPair keyPair = KeyPair.fromAccountId(accountId);
+    byte[] txHash = transaction.hash();
+    for (DecoratedSignature decoratedSignature : transaction.getSignatures()) {
+      if (Arrays.equals(decoratedSignature.getHint().getSignatureHint(), keyPair.getSignatureHint().getSignatureHint()) && keyPair.verify(txHash, decoratedSignature.getSignature().getSignature())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Used to store the results produced by {@link Sep10Challenge#readChallengeTransaction(String, String, Network)}.
+   */
+  public static class ChallengeTransaction {
+    private Transaction transaction;
+    private String clientAccountId;
+
+    public ChallengeTransaction(Transaction transaction, String clientAccountId) {
+      this.transaction = transaction;
+      this.clientAccountId = clientAccountId;
+    }
+
+    public Transaction getTransaction() {
+      return transaction;
+    }
+
+    public void setTransaction(Transaction transaction) {
+      this.transaction = transaction;
+    }
+
+    public String getClientAccountId() {
+      return clientAccountId;
+    }
+
+    public void setClientAccountId(String clientAccountId) {
+      this.clientAccountId = clientAccountId;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(this.transaction, this.clientAccountId);
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object == this) {
+        return true;
+      }
+
+      if (!(object instanceof ChallengeTransaction)) {
+        return false;
+      }
+
+      ChallengeTransaction other = (ChallengeTransaction) object;
+      return Objects.equal(this.transaction, other.transaction) &&
+              Objects.equal(this.clientAccountId, other.clientAccountId);
+    }
+  }
+
+  /**
+   * Represents signers.
+   */
+  public static class Signer {
+    private String key;
```

Comment:
it seems like `Signer ` should be immutable. I would make these fields final and remove their setters